File: backend/app/api/routes/stages.py
"""Stage workspace preparation and G07 approval endpoints."""

from fastapi import APIRouter, Depends, HTTPException

from app.api.stage_contracts import (
    G07DecisionRequest,
    G07ReviewResponse,
    StageBootstrapInstallRequest,
    StageBootstrapInstallResponse,
    StageBootstrapStatusResponse,
    StagePrepareRequest,
    StagePrepareResponse,
    StageSandboxRequest,
    StageSandboxResponse,
)
from app.services.stage_preparation_service import StageApplicationError, StagePreparationApplicationService
from app.services.stage_bootstrap_service import StageBootstrapApplicationService

router = APIRouter(prefix="/runs", tags=["stages"])


def get_stage_service() -> StagePreparationApplicationService:
    return StagePreparationApplicationService()


def get_bootstrap_service() -> StageBootstrapApplicationService:
    return StageBootstrapApplicationService()


@router.post("/{run_id}/stages/prepare", response_model=StagePrepareResponse)
def prepare_stage(run_id: str, request: StagePrepareRequest, service: StagePreparationApplicationService = Depends(get_stage_service)):
    try:
        return service.prepare_stage(run_id, request)
    except StageApplicationError as e:
        raise HTTPException(status_code=e.status_code, detail={"error_code": e.code, "message": e.message}) from e


@router.post("/{run_id}/stages/{stage_id}/sandbox", response_model=StageSandboxResponse)
def create_sandbox(run_id: str, stage_id: str, request: StageSandboxRequest, service: StagePreparationApplicationService = Depends(get_stage_service)):
    try:
        return service.create_sandbox(run_id, stage_id, request)
    except StageApplicationError as e:
        raise HTTPException(status_code=e.status_code, detail={"error_code": e.code, "message": e.message}) from e


@router.get("/{run_id}/approvals/G07", response_model=G07ReviewResponse)
def inspect_g07(run_id: str, stage_id: str | None = None, service: StagePreparationApplicationService = Depends(get_stage_service)):
    """Get G07 gate status. Requires stage_id query parameter."""
    if not stage_id:
        raise HTTPException(status_code=400, detail="stage_id query parameter is required")
    result = service.get_g07(run_id, stage_id)
    if result is None:
        raise HTTPException(status_code=404, detail="G07 approval package not found")
    return result


@router.post("/{run_id}/approvals/G07/decisions", response_model=G07ReviewResponse)
def decide_g07(run_id: str, request: G07DecisionRequest, service: StagePreparationApplicationService = Depends(get_stage_service)):
    if request.gate_id != "G07":
        raise HTTPException(status_code=400, detail="gate_id in body does not match path")
    if not request.stage_id:
        raise HTTPException(status_code=400, detail="stage_id is required in the request body")
    try:
        return service.decide_g07(run_id, request.stage_id, request)
    except StageApplicationError as e:
        raise HTTPException(status_code=e.status_code, detail={"error_code": e.code, "message": e.message}) from e


@router.post("/{run_id}/stages/{stage_id}/bootstrap-install", response_model=StageBootstrapInstallResponse)
def bootstrap_install(run_id: str, stage_id: str, request: StageBootstrapInstallRequest, service: StageBootstrapApplicationService = Depends(get_bootstrap_service)):
    try:
        return service.run_bootstrap_install(run_id, stage_id, request)
    except StageApplicationError as e:
        raise HTTPException(status_code=e.status_code, detail={"error_code": e.code, "message": e.message}) from e


@router.get("/{run_id}/stages/{stage_id}/steps/bootstrap-install", response_model=StageBootstrapStatusResponse)
def get_bootstrap_status(run_id: str, stage_id: str, service: StageBootstrapApplicationService = Depends(get_bootstrap_service)):
    result = service.get_bootstrap_status(run_id, stage_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Bootstrap install step not found")
    return result

File: frontend/src/api/stages.ts
import { apiClient, type createApiClient } from "./client";

export interface StagePrepareRequest {
  expected_state_version: number;
  idempotency_key: string;
  actor: string;
  stage_key: string;
  source_version_family: string;
  target_version_family: string;
  plan_version: string;
}

export interface StagePrepareResponse {
  run_id: string;
  stage_id: string;
  stage_key: string;
  status: string;
  state_version: number;
  event_sequence: number;
  plan: Record<string, unknown> | null;
  idempotent_replay: boolean;
}

export interface G07DecisionRequest {
  expected_state_version: number;
  idempotency_key: string;
  actor: string;
  stage_id: string;
  decision: "approved" | "approved_with_comment" | "modification_requested" | "rejected";
  comment?: string | null;
  gate_id?: string;
}

export interface G07ReviewResponse {
  run_id: string;
  stage_id: string;
  gate_id: string;
  gate_version: string;
  status: string;
  decision: string | null;
  package: Record<string, unknown>;
  state_version: number;
  event_sequence: number;
  idempotent_replay: boolean;
  stale_reason: string | null;
  comment: string | null;
  decision_id?: string | null;
}

export interface StageSandboxRequest {
  expected_state_version: number;
  idempotency_key: string;
  actor: string;
}

export interface StageSandboxResponse {
  run_id: string;
  stage_id: string;
  sandbox_path: string;
  status: string;
  state_version: number;
  event_sequence: number;
  verification: Record<string, unknown> | null;
  idempotent_replay: boolean;
}

export interface StageBootstrapInstallRequest {
  expected_state_version: number;
  idempotency_key: string;
  actor: string;
  profile?: string | null;
}

export interface StageBootstrapInstallResponse {
  run_id: string;
  stage_id: string;
  step_id: string;
  status: string;
  command: string | null;
  exit_code: number | null;
  started_at: string | null;
  completed_at: string | null;
  state_version: number;
  event_sequence: number;
  artifact_ids: string[];
  idempotent_replay: boolean;
}

export interface StageBootstrapStatusResponse {
  run_id: string;
  stage_id: string;
  step_id: string;
  name: string;
  status: string;
  command: string | null;
  exit_code: number | null;
  started_at: string | null;
  completed_at: string | null;
  artifact_ids: string[];
}

type ApiClient = ReturnType<typeof createApiClient>;

export function prepareStage(
  runId: string,
  stageId: string,
  request: StagePrepareRequest,
  client: ApiClient = apiClient,
): Promise<StagePrepareResponse> {
  // Note: stage_key is in the request body; the backend route is /stages/prepare (no stage_id in URL)
  return client.post<StagePrepareResponse>(
    `/api/v1/runs/${encodeURIComponent(runId)}/stages/prepare`,
    request,
  );
}

export function createSandbox(
  runId: string,
  stageId: string,
  request: StageSandboxRequest,
  client: ApiClient = apiClient,
): Promise<StageSandboxResponse> {
  return client.post<StageSandboxResponse>(
    `/api/v1/runs/${encodeURIComponent(runId)}/stages/${encodeURIComponent(stageId)}/sandbox`,
    request,
  );
}

export function getG07Status(
  runId: string,
  stageId: string,
  client: ApiClient = apiClient,
): Promise<G07ReviewResponse> {
  const params = new URLSearchParams({ stage_id: stageId });
  return client.get<G07ReviewResponse>(
    `/api/v1/runs/${encodeURIComponent(runId)}/approvals/G07?${params.toString()}`,
  );
}

export function decideG07(
  runId: string,
  request: G07DecisionRequest,
  client: ApiClient = apiClient,
): Promise<G07ReviewResponse> {
  return client.post<G07ReviewResponse>(
    `/api/v1/runs/${encodeURIComponent(runId)}/approvals/G07/decisions`,
    request,
  );
}

export function runBootstrapInstall(
  runId: string,
  stageId: string,
  request: StageBootstrapInstallRequest,
  client: ApiClient = apiClient,
): Promise<StageBootstrapInstallResponse> {
  return client.post<StageBootstrapInstallResponse>(
    `/api/v1/runs/${encodeURIComponent(runId)}/stages/${encodeURIComponent(stageId)}/bootstrap-install`,
    request,
  );
}

export function getBootstrapInstallStatus(
  runId: string,
  stageId: string,
  client: ApiClient = apiClient,
): Promise<StageBootstrapStatusResponse> {
  return client.get<StageBootstrapStatusResponse>(
    `/api/v1/runs/${encodeURIComponent(runId)}/stages/${encodeURIComponent(stageId)}/steps/bootstrap-install`,
  );
}
